satellite/gracefulexit/endpoint.go

```diff
@@ -0,0 +1,453 @@
+// Copyright (C) 2019 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package gracefulexit
+
+import (
+	"context"
+	"io"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"go.uber.org/zap"
+	"golang.org/x/sync/errgroup"
+
+	"storj.io/storj/pkg/identity"
+	"storj.io/storj/pkg/pb"
+	"storj.io/storj/pkg/rpc/rpcstatus"
+	"storj.io/storj/pkg/storj"
+	"storj.io/storj/satellite/metainfo"
+	"storj.io/storj/satellite/orders"
+	"storj.io/storj/satellite/overlay"
+	"storj.io/storj/uplink/eestream"
+)
+
+// millis for the transfer queue building ticker
+const buildQueueMillis = 100
+
+// drpcEndpoint wraps streaming methods so that they can be used with drpc
+type drpcEndpoint struct{ *Endpoint }
+
+// processStream is the minimum interface required to process requests.
+type processStream interface {
+	Context() context.Context
+	Send(*pb.SatelliteMessage) error
+	Recv() (*pb.StorageNodeMessage, error)
+}
+
+// Endpoint for handling the transfer of pieces for Graceful Exit.
+type Endpoint struct {
+	log       *zap.Logger
+	interval  time.Duration
+	db        DB
+	overlaydb overlay.DB
+	overlay   *overlay.Service
+	metainfo  *metainfo.Service
+	orders    *orders.Service
+	config    Config
+}
+
+type pendingTransfer struct {
+	path             []byte
+	pieceSize        int64
+	satelliteMessage *pb.SatelliteMessage
+}
+
+// pendingMap for managing concurrent access to the pending transfer map.
+type pendingMap struct {
+	mu   sync.RWMutex
+	data map[storj.PieceID]*pendingTransfer
+}
+
+// newPendingMap creates a new pendingMap and instantiates the map.
+func newPendingMap() *pendingMap {
+	newData := make(map[storj.PieceID]*pendingTransfer)
+	return &pendingMap{
+		data: newData,
+	}
+}
+
+// put adds to the map.
+func (pm *pendingMap) put(pieceID storj.PieceID, pendingTransfer *pendingTransfer) {
+	pm.mu.Lock()
+	defer pm.mu.Unlock()
+
+	pm.data[pieceID] = pendingTransfer
+}
+
+// get returns the pending transfer item from the map, if it exists.
+func (pm *pendingMap) get(pieceID storj.PieceID) (pendingTransfer *pendingTransfer, ok bool) {
+	pm.mu.RLock()
```

Comment:
is locking the mutex necessary for `get`? I thought maps were safe for concurrent reads.

Reply:
The locks on `get` are necessary.  Otherwise, wonkiness can happen.
see https://blog.golang.org/go-maps-in-action